### PR TITLE
numa_opts: add new case to test 128 numa nodes

### PR DESCRIPTION
**Pull request by huangyum**
ID: 1666144
Signed-off-by: Yumei Huang <yuhuang@redhat.com>

### CODE REVIEW

**Comment by huangyum** [comment]
Test passed on x86 host. @MiriamDeng Please help review and test it on ppc host, thanks!
(1/1) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.numa_opts.nodes.128: PASS (174.68 s)

File: qemu/tests/cfg/numa_opts.cfg [hunk 1]
```diff
@@ -92,3 +92,19 @@
               extra_params += " -numa node,mem=2560,cpus=5,nodeid=2"
               extra_params += " -numa node,mem=512,cpus=3-4,nodeid=0"
               extra_params += " -numa node,mem=1024,cpus=0-2,nodeid=1"
+    - nodes.128:
+        type = numa_maxnodes
+        host_node = 0
+        node_num = 128
```
Comment (yanan-fu) [inline qemu/tests/cfg/numa_opts.cfg]: For other variants, `nodes.1`, `nodes.2`...   it use `numa_nodes = *` ,  but here it use `node_num = 128`.
Could you explain the difference between `numa_nodes` and `node_num`  ?

Comment (huangyum) [reply qemu/tests/cfg/numa_opts.cfg]: Seems they have same meanning, have updated, thanks.

**Comment by MiriamDeng** [comment]
Hi huangyum,
    I just want to confirm with you about two questions before I proceed it further
    1.if the default node  can't support 128's node's memory,can it get the rest memory from other nodes ?
    2.if all available nodes can't support,any possibilities for us to support less 128 node test.
Thanks.

**Comment by MiriamDeng** [comment]
Hi huangyum,
I just want to confirm with you about two questions before I proceed it further
1.if the default node can't support 128's node's memory,can it get the rest memory from other nodes ?
2.if all available nodes can't support it,is there any possibilities for us to support less 128 nodes test.
Thanks.

**Comment by huangyum** [comment]
@MiriamDeng With bind policy, it can not get memory from other nodes.  The auto test steps are consistent with the test case steps. I see your qemu cli for ppc.  If the ppc host node doesn't have enough memory to boot guest,  we may change the bind policy to other ones, what do you think?

**Comment by huangyum** [comment]
@MiriamDeng The case is intended to test with max number of numa nodes supported by qemu, so I don't think we should reduce the number. How about we change the policy to default, so as long as host has enough memory, guest could boot successfully.

**Comment by MiriamDeng** [comment]
@huangyum,got your points and it's acceptable for me.Thanks.

**Comment by huangyum** [comment]
@MiriamDeng Have updated to default policy, test passed on x86. Please help review.

**Comment by MiriamDeng** [comment]
Job could passed on ppc,thanks.
 (1/1) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.numa_opts.nodes.128: PASS (94.23 s)

**Comment by huangyum** [comment]
@MiriamDeng  Thanks for your test.
@yanan-fu would you please help review again, thanks.

**Comment by huangyum** [comment]
@MiriamDeng Does the case fail due to host lack of memory or other reasons?   When you do the manual test, what do you do when host doesn't have enough memory? If you set prealloc=no, you are actually using memory overcommit, which has limitation and may hit error after guest boot up.

File: qemu/tests/numa_maxnodes.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        params["size_mem%s" % index] = node_size
+        params["prealloc_mem%s" % index] = prealloc_mem
+
+    params["guest_numa_nodes"] = guest_numa_nodes
+    params["mem_devs"] = mem_devs
+    params["start_vm"] = "yes"
+    env_process.preprocess_vm(test, params, env, params["main_vm"])
+    error_context.context("Get the main VM!", logging.info)
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
```
Comment (vivianQizhu) [inline qemu/tests/numa_maxnodes.py]: This is unnecessary since you perform `wait_for_login()` immediately.

File: qemu/tests/cfg/numa_opts.cfg [hunk 3]
```diff
@@ -92,3 +92,18 @@
               extra_params += " -numa node,mem=2560,cpus=5,nodeid=2"
               extra_params += " -numa node,mem=512,cpus=3-4,nodeid=0"
               extra_params += " -numa node,mem=1024,cpus=0-2,nodeid=1"
+    - nodes.128:
+        type = numa_maxnodes
+        numa_nodes = 128
+        mem_fixed = 4G
+        use_mem = no
+        node_size = 32M
```
Comment (vivianQizhu) [inline qemu/tests/cfg/numa_opts.cfg]: Better define `start_vm = no`, otherwise I think the vm is processed twice.

Comment (huangyum) [reply qemu/tests/cfg/numa_opts.cfg]: It has been defined in L5.

File: qemu/tests/numa_maxnodes.py (hunk 2 above)
Comment (huangyum) [reply qemu/tests/numa_maxnodes.py]: Thanks, updated.

**Comment by MiriamDeng** [comment]
@huangyum,if the host's memory is not enough,the test case will fail.For manual test,we need to find enough memory host or change option accordingly if there's no hardware available.Checking twice it should work on ppc.
     By the way,how do you think if we execute something like "echo 3 > /proc/sys/vm/drop_caches" before running this job.Thanks.

**Comment by huangyum** [comment]
@MiriamDeng Thanks for your test.  It's a good idea to drop caches before test. Have updated, thanks.

File: qemu/tests/cfg/numa_opts.cfg [hunk 4]
```diff
@@ -8,6 +8,7 @@
   kill_vm = yes
   kill_vm_gracefully = no
   encode_video_files = no
+  pre_command = "sync && echo 3 >/proc/sys/vm/drop_caches"
```
Comment (vivianQizhu) [inline qemu/tests/cfg/numa_opts.cfg]: Or we might use `utils_memory.drop_caches()`

**Comment by vivianQizhu** [comment]
@huangyum Are we able to evaluate the test environment at the beginning of the test, and cancel it if host capacity not sufficient, and also notify user to find out proper environment for the test?

File: qemu/tests/cfg/numa_opts.cfg (hunk 4 above)
Comment (huangyum) [reply qemu/tests/cfg/numa_opts.cfg]: Yes, we could, but then we need update in several py. I would prefer to keep it is, thanks.

**Comment by huangyum** [comment]
@vivianQizhu I set vm_mem_minimum for ppc since it needs 32G memory. Case would get canceled when host does not have enough memory. Thanks.

**Comment by vivianQizhu** [comment]
@huangyum Ah there it is, then it looks good to me.

File: qemu/tests/cfg/numa_opts.cfg (hunk 4 above)
Comment (yanan-fu) [inline qemu/tests/cfg/numa_opts.cfg]: Better append `;` at the end the pre_command.

Comment (huangyum) [reply qemu/tests/cfg/numa_opts.cfg]: Thanks, updated.

**Comment by MiriamDeng** [comment]
LGTM,thanks

File: qemu/tests/numa_maxnodes.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+from virttest import env_process
+from virttest import error_context
+from qemu.tests import numa_memdev_options
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    [Memory][Numa]binding 128 guest nodes to node0 - x86 witn 32M and Power
```
Comment (yanan-fu) [inline qemu/tests/numa_maxnodes.py]: typo, `with`

File: qemu/tests/cfg/numa_opts.cfg [hunk 6]
```diff
@@ -92,3 +93,19 @@
...
+        mem_fixed = 4G
+        use_mem = no
+        node_size = 32M
+        ppc64,ppc64le:
+            mem_fixed = 32G
+            node_size = 256M
+            vm_mem_minimum = 32G
+        paused_after_start_vm = no
+        prealloc_mem = yes
+        policy_mem = default
```
Comment (yanan-fu) [inline qemu/tests/cfg/numa_opts.cfg]: I am not familiar with this feature, but in polarion, both x86 and ppc use `policy=bind`,  just want to double confirm with you.

File: qemu/tests/cfg/numa_opts.cfg [hunk 7]
```diff
@@ -92,3 +93,19 @@
               extra_params += " -numa node,mem=2560,cpus=5,nodeid=2"
               extra_params += " -numa node,mem=512,cpus=3-4,nodeid=0"
               extra_params += " -numa node,mem=1024,cpus=0-2,nodeid=1"
+    - nodes.128:
+        type = numa_maxnodes
+        numa_nodes = 128
+        mem_fixed = 4G
+        use_mem = no
+        node_size = 32M
```
Comment (yanan-fu) [inline qemu/tests/cfg/numa_opts.cfg]: In my opinion, better add  `vm_mem_minimum=4G`  for x86 (32M*128) although low chance to get this kind of machine. what do you think ?

File: qemu/tests/numa_maxnodes.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    3) Check memory in procfs on host
+    4) Check numa node amount in guest
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+    error_context.context("Modify params to boot guest with 128 numa nodes",
+                          logging.info)
+    node_num = int(params.get("numa_nodes", 1))
```
Comment (yanan-fu) [inline qemu/tests/numa_maxnodes.py]: prefer `params["numa_nodes"]` since this case has fixed numa_nodes.

File: qemu/tests/numa_maxnodes.py (hunk 5 above)
Comment (yanan-fu) [inline qemu/tests/numa_maxnodes.py]: typo, `with`

File: qemu/tests/cfg/numa_opts.cfg (hunk 6 above)
Comment (huangyum) [reply qemu/tests/cfg/numa_opts.cfg]: @yanan-fu This case is to test 128 nodes, either policy is ok. And if policy=bind, we need the specified host node has all the memory, which could be hard to find such host for ppc.  So we decided to use default policy instead. Thanks.

File: qemu/tests/cfg/numa_opts.cfg (hunk 7 above)
Comment (huangyum) [reply qemu/tests/cfg/numa_opts.cfg]: @yanan-fu I've thought about it. Considering we normally run memory test on host with more than 8G memory and never hit the issue, I didn't add the param for x86. But I guess it should be better to treat x86 and ppc equally, will update, thanks.

File: qemu/tests/numa_maxnodes.py (hunk 5 above)
Comment (huangyum) [reply qemu/tests/numa_maxnodes.py]: Updated, plese help review again, thanks.

File: qemu/tests/cfg/numa_opts.cfg (hunk 6 above)
Comment (yanan-fu) [reply qemu/tests/cfg/numa_opts.cfg]: @huangyum  we don't specify `host-nodes=0` as the polarion case, reason is same with above, right ?

Comment (huangyum) [reply qemu/tests/cfg/numa_opts.cfg]: @yanan-fu Yes, we can not set host node when using default policy.

**Comment by yanan-fu** [comment]
code LGTM, and test pass with latest code for both rhel and windows guest:
 (1/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.io-github-autotest-qemu.numa_opts.nodes.128: PASS (68.27 s)
 (2/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.numa_opts.nodes.128: PASS (120.92 s)

Acked-by: Yanan Fu <yfu@redhat.com>

**Pull request merged**